### ISSUE TITLE
different learning rate

### ISSUE THREAD

**Issue by guangyuli-uoe** [Lightning-AI/litgpt#2146]
hi,

i would like to use different learning rate for different layer,
currently im using the pretrain.py,

but i have no idea how to do it in litgpt,

looking forward to your kind reply,

**Comment by Andrei-Aksionov** [comment]
Hello @guangyuli-uoe 

TBH never tried this, but I think instead of providing `model.parameters()` to an optimizer:

https://github.com/Lightning-AI/litgpt/blob/22c2a4f3d80599b0791d62c4c4fb871bb5486609/litgpt/pretrain.py#L217

you can create a list of dictionaries, where you can assign a learning rate, that is different to the default one, and do it for each layer.

Here is a simple code snippet with an example:

```python
import torch
import torch.nn as nn
import torch.optim as optim

# Define a simple CNN model
class MyCNN(nn.Module):
    def __init__(self):
        super(MyCNN, self).__init__()
        self.conv_layers = nn.Sequential(
            nn.Conv2d(1, 16, kernel_size=3, padding=1),
            nn.ReLU(),
            nn.MaxPool2d(2),
            nn.Conv2d(16, 32, kernel_size=3, padding=1),
            nn.ReLU(),
            nn.MaxPool2d(2)
        )
        self.fc_layers = nn.Sequential(
            nn.Linear(32 * 7 * 7, 128),
            nn.ReLU(),
            nn.Linear(128, 10)
        )
        self.cls = nn.Linear(10, 1)

    def forward(self, x):
        x = self.conv_layers(x)
        x = x.view(x.size(0), -1)
        x = self.fc_layers(x)
        x = self.cls(x)
        return x

# Instantiate the model
model = MyCNN()

# Separate the parameters into groups
conv_params = list(model.conv_layers.parameters())
fc_params = list(model.fc_layers.parameters())

# Create a set of parameters that have custom learning rates
custom_lr_params = set(conv_params + fc_params)

# Create the list of parameter groups
param_groups = [
    {'params': conv_params, 'lr': 3e-4, 'name': 'convolutional_group'},
    {'params': fc_params, 'lr': 5e-2, 'name': 'fully_connected_group'}
]

# Add the remaining parameters with the default learning rate
param_groups.append({'params': [p for p in model.parameters() if p not in custom_lr_params], 'name': 'default_group'})


# Initialize the optimizer with the parameter groups
optimizer = optim.Adam(param_groups, lr=1e-3)

# Print optimizer's parameter groups, now accessing the 'name'
print("Optimizer's parameter groups:")
for i, group in enumerate(optimizer.param_groups):
    print(f"Group {i} (Name: {group['name']}):")
    print(f"  Learning Rate: {group['lr']}")
    print(f"  Number of params: {len(group['params'])}")
```

and the output will be:
```text
Optimizer's parameter groups:
Group 0 (Name: convolutional_group):
  Learning Rate: 0.0003
  Number of params: 4
Group 1 (Name: fully_connected_group):
  Learning Rate: 0.05
  Number of params: 4
Group 2 (Name: default_group):
  Learning Rate: 0.001
  Number of params: 2
```
----

From the example you can see, that for the last layer, for which I didn't specify a custom learning rate, the default one (the one that was provided into an optimizer class init) was used.
Of course, you can make it more granular, i.e. assign a custom LR per each layer in self.conv_layer.

**Comment by guangyuli-uoe** [comment]
@Andrei-Aksionov 
hi ! 
really thanks for your kind reply, and the code you provided is helpful !!! 

i also want to know if the current litgpt support gpt2 series model (small, medium, ?? cause they use abs position ids TT


best wishes

**Comment by bhimrazy** [comment]
Hi @guangyuli-uoe , it seems litgpt does not currently support GPT-2 series models.

**Comment by bhimrazy** [comment]
Closing this since it’s resolved — please feel free to reopen if you run into any further issues.
